### ISSUE TITLE
Categorization (personal, business, services)

### ISSUE THREAD

**Issue by joebew42** [joaquimadraz/opensubs.io#24]
Hi @joaquimadraz ,

I am doing some live coding on [Twitch](http://twitch.tv/joebew42) and I would like to contribute to your project, so I tried to take a look and I've seen that there is something in the "nice to have" features that I could try to tackle: the categorization (personal, business, services) feature.

I would like to know how do you imagine that feature is supposed to be implemented. If you provide some enough instructions I think I can start work on it. In the meantime I'll try to setup the project on my local machine and undestand something about it.

Thank you,
joe.

**Comment by joaquimadraz** [comment]
Hey @joebew42 thanks for reaching out. I'd be pleased to have another contributor 👍

So my initial goal with the categorization was to have an idea of how the expenses are distributed. 
For instance:

<img width="469" alt="screen shot 2018-05-21 at 21 59 39" src="https://user-images.githubusercontent.com/822817/40329706-4f12330a-5d42-11e8-8c68-70720971d7ce.png">

These stats could be put on the "All payments" page, and later this could also be used to personalize the messages on emails and monthly pagination.

What do you think?

**Comment by joebew42** [comment]
Hey, perfect! In the meantime I'm trying to get the application up and running. I see that there are a few enhancements we can add in order to help people with the onboarding process (setup, seed, etc ...) for the development.

As soon as I am ready I will back to you to ask further details about this feature, ok? Do you have a gitter channel for this project? A place where we can chat about it could be useful. What do you think?

**Comment by joebew42** [comment]
Hi @joaquimadraz ,

I just had the time to run the application on my local machine and play with it a little bit.
I am here to ask you some questions.

1. How it is supposed a user should categorize a payment? We have to consider to add a new field `category` for each payment entry? Like the simple mock up that you can find below?

![screenshot-20180601-19 21 33](https://user-images.githubusercontent.com/1238549/40854535-1b9a7a62-65d2-11e8-9930-984a4d38699d.png)

If yes, where we can start with? Can we consider to add a free editable input text box? Or, we may have to consider to have a separate page, like `settings -> categories` where the user can organize all categories and then choose one of it through a dropdown menu' from the `create payment` modal window?

2. Can we associate more than one category to a payment? 

3. How do you think the page `all payments` should be changed? Right now we have this kind of visualization:

![screenshot-20180601-19 21 08](https://user-images.githubusercontent.com/1238549/40854915-3a3d9fde-65d3-11e8-936d-935ac20101fe.png)

The user will switch to the "categorization visualization" by clicking on a link above the list, or do you have in mind some different mechanism?

4. What the `percentage` value will represent there? The number of payments for each category, or the overall expenses for each category?

I think it could be useful if you share some mockups and some more details about this feature.

Thank you again,
joe!

**Comment by joebew42** [comment]
Thank you for all your answers, so clear!

**Comment by joebew42** [comment]
Hi @joaquimadraz,

just another question related to the `Other` category. What happens when the user don't specify a category? We may have to fall back to a default one (e.g. Other) or we may allow user to specify an empty category? In this case, how it is supposed to show the categorization for all the payments that haven't a category? :thinking:

**Comment by joaquimadraz** [comment]
I think it would be interesting to separate `Other` and `Uncategorized` payments.
Here are some things that comes to mind:

- All existing payments (prior to the release of this feature) that are not categorized are placed in the `Uncategorized` category.
- `Other` is something we don't have a category for.
- New payments require a category. 

Does it make sense?

**Comment by joebew42** [comment]
It makes sense. So, we can end up with these requirements:

- All the payments - prior to the release of this feature - will be migrated to a `Uncategorized` category (we translate this by adding the value `uncategorized` in the `category` field?).
- `Other` will be the default value of `category` for all the new payments.
- The field `category` is mandatory and cannot be blank.

Right?

**Comment by joebew42** [comment]
Hi @joaquimadraz ,

I have in mind to end up with this API for the spendings summary, let me know what you think about:

```
HTTP GET /api/subscriptions/summary

{
  data: {
    currency: "EUR",
    currency_symbol: "€",
    spendings: [
      {category: "travel", amount: 239.99},
      {category: "videogames", amount: 999.99},
      {category: "other", amount: 9.99},
      {category: "uncategorized", amount: 1900.00},
    ]
  }
}
```

So, basically the fields `currency` and `currency_symbol` comes from the `User`, and the `spendings` will be a list of all expences grouped by categories. In this way we let the client (e.g. Reactjs) to decide what to do with this data. What do you think?

**Comment by joaquimadraz** [comment]
Hey @joebew42, sorry for the late reply. 

Looks good. Just one thing, should we also return the total spending in this request to calculate the percentages?

**Comment by joebew42** [comment]
> Looks good. Just one thing, should we also return the total spending in this request to calculate the percentages?

Of course, we can adapt the API in order to return the total spending too!
